### PR TITLE
MAINT: Optimize and modularize tests

### LINKED ISSUE #167: clean up unit tests
**Improvement Description**
The unit tests for this plugin are overgrown. It is time to get organized.

**Current Behavior**
At the time of writing, the unit tests are 1546 lines long. Some cleanup of the code may also improve test runtime.

**Proposed Behavior**
1. split up test_classifier.py into multiple thematic test modules. E.g., types/formats should be tested separately. Visualizations and utilities could also be tested separately.
2. split up or combine some test classes, based on shared test data
3. clean up the test data to simplify tests where possible. E.g., the full-sized datasets could be replaced by toy datasets for more tests
    a. `TestHeatmap` should use toy data, the tests take close to 2 minutes to run for a fairly trivial set of tests.

### PR DESCRIPTION
**Pull request by Oddant1**
Partially addresses #167 by modularizing tests and optimizing visualizer tests.

### CODE REVIEW

**Comment by nbokulich** [comment]
Awesome, thanks for getting started on this @Oddant1 !

A major goal of #167 is to use toy datasets for many or most tests instead of the full-sized real datasets (see note 3 in #167). I have only skimmed over your diff, but it looks like that change still needs to occur. I am fine splitting that into a separate PR — but if you want to go that way let's just update #167 instead of closing.

**Comment by Oddant1** [comment]
@nbokulich I intend to update the datasets used in future commits to this PR once the tests are fully modularized

**Comment by nbokulich** [comment]
Awesome, thanks! Just wanted to make sure — could you please mark this as a draft PR and add the "do not merge" label until it is ready for review?

File: q2_sample_classifier/tests/data/importance_chardonney.tsv [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+4393057	1.2376844272339594e-05
+4447334	1.1787784859310052e-05
+845178	1.0706581367368972e-05
+544313	1.0371458403674174e-05
+555725	1.0085471004943226e-05
+13763	8.9648631155042e-06
+4419276	8.951412089688543e-06
+100497	6.6916133546051555e-06
+826250	6.6913819103412555e-06
+
```
Comment (thermokarst) [inline q2_sample_classifier/tests/data/importance_chardonney.tsv]: Drive by comment: Is this file supposed to have all these blank lines in it? If it does we should address this in another PR.

Comment (Oddant1) [reply q2_sample_classifier/tests/data/importance_chardonney.tsv]: No, I didn't notice those were there :laughing: they came from how I added the second column to the file.

Comment (Oddant1) [reply q2_sample_classifier/tests/data/importance_chardonney.tsv]: Something happened and my comment here seems to be gone? But adding it back for posterity. No those lines were not supposed to be there. They came about as a result of how I added the second column to the file and I didn't notice that they were still there.

**Comment by Oddant1** [comment]
I changed the heatmap test to use the chardonnay dataset instead of the dataset it was using, and it cut the time to run all the tests in half on my machine

**Comment by nbokulich** [comment]
what was the dataset used before? the heatmap is one where a toy dataset
should be fine to use, it will probably reduce runtime a bit more (because
ironically the heatmap is one of the most time-intensive steps in the unit
tests for this plugin!)

**Comment by Oddant1** [comment]
@nbokulich the dataset being used before was maturity which is much MUCH larger

**Comment by Oddant1** [comment]
As on https://github.com/qiime2/q2-sample-classifier/pull/188/commits/2daba08946e402888c2ef9a5611e6984dc5e5ac9 the tests are taking 75 - 80 seconds to run on my maching, and some of the tests in `test_estimators.py` seem to be taking the longest of any individual test

**Comment by nbokulich** [comment]
That's because some of the tests in test_estimators.py are looping through
all possible estimators, so what looks like a single test is actually
around a dozen. Those tests in particular would benefit from a small toy
dataset (if you haven't already) but ~75 sec is already a big improvement!

**Comment by Oddant1** [comment]
Yeah, converting those tests over to using a toy dataset is proving more challenging than converting the heatmap tests

**Comment by Oddant1** [comment]
I'm not really sure how to go about creating a toy dataset for the estimator tests. The small dataset I found in diversity and started using for the visualizers seems ill suited for the estimators.

**Comment by Oddant1** [comment]
This is probably the best I'm going to be able to do. The tests are split across multiple files, and the time to run the tests on my machine is down from ~145 seconds to ~65 seconds

**Comment by Oddant1** [comment]
@nbokulich, looking at cleaning this up now. The tests in `test_estimators.py` were mostly changed from maturity to chardonnay where possible to further reduce runtime of tests because chardonnay is much smaller than maturity.

File: q2_sample_classifier/tests/test_estimators.py [hunk 2]
```diff
@@ -0,0 +1,605 @@
...
+    def _custom_setup(self, version):
+        with open(os.path.join(self.temp_dir.name,
+                               'sklearn_version.json'), 'w') as fh:
+            fh.write(json.dumps({'sklearn-version': version}))
+        shutil.copy(self.sklearn_pipeline, self.temp_dir.name)
+        return SampleEstimatorDirFmt(
+            self.temp_dir.name, mode='r')
+
+
+class TestFormats(SampleEstimatorTestBase):
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_estimators.py]: move to separate module?

File: q2_sample_classifier/tests/test_estimators.py [hunk 3]
```diff
@@ -0,0 +1,605 @@
...
+    def test_sample_classifier_dir_fmt(self):
+        format = self._custom_setup(sklearn.__version__)
+
+        # Should not error
+        format.validate()
+
+
+class TestTransformers(SampleEstimatorTestBase):
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_estimators.py]: move to separate module?

File: q2_sample_classifier/tests/test_estimators.py [hunk 4]
```diff
@@ -0,0 +1,605 @@
...
+        transformer = self.get_transformer(
+            Pipeline, SampleEstimatorDirFmt)
+        obs = transformer(exp)
+        sklearn_pipeline = obs.sklearn_pipeline.view(PickleFormat)
+        obs_pipeline = read_pipeline(str(sklearn_pipeline))
+        obs = obs_pipeline
+        self.assertTrue(obs)
+
+
+# make sure summarize visualizer works and that rfe_scores are stored properly
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_estimators.py]: move to the test_actions module?

File: q2_sample_classifier/tests/test_estimators.py [hunk 5]
```diff
@@ -0,0 +1,605 @@
...
+        def _load_cmc(md_fp, column):
+            md_fp = self.get_data_path(md_fp)
+            md = pd.read_csv(md_fp, sep='\t', header=0, index_col=0)
+            md = qiime2.CategoricalMetadataColumn(md[column])
+            return md
+
+        self.table_chard_fp = _load_biom('chardonnay.table.qza')
+        self.md_chard_fp = _load_md('chardonnay.map.txt')
+        self.mdc_chard_fp = _load_cmc('chardonnay.map.txt', 'Region')
+        self.mdc_chard_num_fp = _load_nmc('chardonnay.map.txt', 'Vineyard')
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_estimators.py]: even though vineyard is represented by a number, this is categorical data and not an appropriate source of test data for this test.

**Comment by Oddant1** [comment]
@nbokulich I think I cleaned up everything mentioned in the comments. Test runtime is at ~70-75 seconds on my machine with the estimators using maturity again.

File: q2_sample_classifier/tests/test_visualization.py [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+        table_vaw = self.get_data_path('vaw.qza')
+        self.table_vaw = qiime2.Artifact.load(table_vaw)
+        self.table_vaw, = feature_table.actions.filter_samples(
+            self.table_vaw, metadata=md_vaw)
+        imp = pd.read_csv(
+            self.get_data_path('vaw_importance.tsv'), sep='\t',
+            header=0, index_col=0)
+        self.imp = qiime2.Artifact.import_data('FeatureData[Importance]', imp)
+
+    def test_heatmap_default_feature_count_zero(self):
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_visualization.py]: is this test still working properly? with feature_count=0 it should be displaying all features... looks like there are 5 features in the input table, but only 4 in the heatmap? Why is one being dropped?

File: q2_sample_classifier/tests/test_visualization.py [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+        heatmap, table, = sample_classifier.actions.heatmap(
+            self.table_vaw, self.imp, self.md_vaw, group_samples=True,
+            feature_count=0)
+        self.assertEqual(table.view(biom.Table).shape, (4, 2))
+
+    def test_heatmap_importance_threshold(self):
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_visualization.py]: this test does not appear to be working correctly anymore, either, probably for the same reason as above... you kept the importance_threshold from the original test, which is lower than all of the new importance scores so should have no effect (choosing a different threshold would be ideal). So I would expect 5 features, but it looks like only 4 are in the heatmap?

File: q2_sample_classifier/tests/test_visualization.py (hunk 6 above)
Comment (Oddant1) [reply q2_sample_classifier/tests/test_visualization.py]: That is a very good question, the answer to which is not immediately obvious to either myself or @ebolyen. Apologies for not noticing something like this earlier. To be perfectly honest my grasp of what these tests really do is shaky at best :laughing:. I'm looking into what might be causing this discrepancy now.

Comment (nbokulich) [reply q2_sample_classifier/tests/test_visualization.py]: see my review comment above for the cause... lack of a header line

Comment (Oddant1) [reply q2_sample_classifier/tests/test_visualization.py]: How did I not see that in your review or in the file 😂. I think it's time for me to be done for the week lol. Been doing too much stuff and getting not enough sleep.

Comment (nbokulich) [reply q2_sample_classifier/tests/test_visualization.py]: lol no worries! I think we are all ready for the weekend! Have a good one!

File: q2_sample_classifier/tests/test_visualization.py [hunk 8]
```diff
@@ -58,8 +58,7 @@ def setUp(self):
         self.md_vaw = md_vaw.get_column('Column')
         table_vaw = self.get_data_path('vaw.qza')
         self.table_vaw = qiime2.Artifact.load(table_vaw)
-        self.table_vaw, = feature_table.actions.filter_samples(
-            self.table_vaw, metadata=md_vaw)
+        self.table_vaw.save('/tmp/vaw')
```
Comment (nbokulich) [inline q2_sample_classifier/tests/test_visualization.py]: why are you saving this table now? This is not clear from the other changes in this commit.

**Comment by nbokulich** [comment]
Thanks @Oddant1 ! You have checked all my boxes, tests are passing, I am merging. Thanks for taking on this herculean and long-overdue task!
